### PR TITLE
Reenable error code and message assertions in scala tests

### PR DESCRIPTION
**Pull request by tomhoule**

### CODE REVIEW

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/UpdateMutationSpec.scala [hunk 1]
```diff
@@ -161,8 +161,8 @@ class UpdateMutationSpec extends FlatSpec with Matchers with ApiSpecBase {
          |  }
          |}""".stripMargin,
       project,
-      errorCode = 3039,
-      errorContains = "No Node for the model Todo with value NOT A VALID ALIAS for alias found"
+      errorCode = 0, // 3039,
+      errorContains = """InterpretationError(\"Error for binding \\'0\\': RecordNotFound(\\\"Record to update not found.\\\")\""""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/UpdateMutationSpec.scala]: Probably not something users can/should handle but we could return a better error message here probably.

File: query-engine/connector-test-kit/src/test/scala/writes/withId/ExecuteRawSpec.scala [hunk 2]
```diff
@@ -210,7 +210,8 @@ class ExecuteRawSpec extends WordSpecLike with Matchers with ApiSpecBase {
          |}
       """.stripMargin,
       project,
-      errorCode = errorCode,
+      errorCode = 2010,
+        // errorCode,
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/withId/ExecuteRawSpec.scala]: we could add an argument to this assertion to check for the error metadata (it's in the JSON we return), that way we could check for the database error code here

File: query-engine/connectors/sql-query-connector/src/database/operations/write.rs [hunk 3]
```diff
@@ -31,6 +32,7 @@ pub async fn create_record(conn: &dyn QueryExt, model: &ModelRef, args: WriteArg
                     let constraint = DatabaseConstraint::Fields(fields.clone());
                     return Err(SqlError::NullConstraintViolation { constraint });
                 }
+                quaint::error::DatabaseConstraint::ForeignKey => unreachable!(),
```
Comment (tomhoule) [inline query-engine/connectors/sql-query-connector/src/database/operations/write.rs]: This variant is on new versions of quaint, I'll wait for another PR to handle it before merging this one, so it's not implemented twice.

File: query-engine/connectors/sql-query-connector/src/error.rs [hunk 4]
```diff
@@ -189,6 +189,7 @@ impl From<quaint::error::Error> for SqlError {
             e @ QuaintKind::ConnectTimeout { .. } => SqlError::ConnectionError(e.into()),
             e @ QuaintKind::Timeout(..) => SqlError::ConnectionError(e.into()),
             e @ QuaintKind::TlsError { .. } => Self::ConnectionError(e.into()),
+            _ => todo!(),
```
Comment (tomhoule) [inline query-engine/connectors/sql-query-connector/src/error.rs]: Same as previous comment.

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteSpec.scala [hunk 5]
```diff
@@ -53,8 +53,9 @@ class DeleteSpec extends FlatSpec with Matchers with ApiSpecBase {
         |}
       """.stripMargin,
       project,
-      errorCode = 3039,
-      errorContains = """No Node for the model Todo with value does not exist for title found."""
+      errorCode = 0, // 3039,
+      errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'0\\': RecordNotFound(\\\"Record to delete does not exist."""
+      // """No Node for the model Todo with value does not exist for title found."""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteSpec.scala]: Should we have a `RecordNotFound` user facing error for this?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala [hunk 6]
```diff
@@ -1165,8 +1166,8 @@ class NestedUpdateMutationInsideUpdateSpec extends FlatSpec with Matchers with A
     server.queryThatMustFail(
       updateMutation,
       project,
-      errorCode = 3041,
-      errorContains = """The relation BottomToMiddle has no node for the model Middle connected to a Node for the model Bottom on your mutation path."""
+      errorCode = 0, // 3041,
+      errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'3\\': AssertionError(\\\"Expected a valid parent ID to be present for nested update to-one case.\\\")\")"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala]: I think this should be an instance of the user facing `MissingRequiredValue` error, wdyt?

File: query-engine/connector-test-kit/src/test/scala/writes/scalarLists/CreateMutationListSpec.scala [hunk 7]
```diff
@@ -90,7 +90,8 @@ class CreateMutationListSpec extends FlatSpec with Matchers with ApiSpecBase {
          |  }){optStrings, optInts, optFloats, optBooleans, optEnums, optDateTimes }
          |}""",
       project = project,
-      errorCode = 1
+      errorCode = 0, // 1
+      errorContains = """Error in query graph construction: InputError(\"The `set` argument was not provided for field `optStrings` on `ScalarModel`"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/scalarLists/CreateMutationListSpec.scala]: I think this should be an instance of the user facing MissingRequiredValue error, wdyt?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala (hunk 6 above)
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala]: For the errors in this file in general: should we try to improve the messages there?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideUpdateSpec.scala [hunk 8]
```diff
@@ -806,8 +807,9 @@ class NestedConnectMutationInsideUpdateSpec extends FlatSpec with Matchers with
            |}
       """,
         project,
-        errorCode = 3042,
-        errorContains = "The change you are trying to make would violate the required relation 'ChildToParent' between Child and Parent"
+        errorCode = 0, // 3042,
+        errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'2\\': RelationViolation(RelationViolation { relation_name: \\\"ChildToParent\\\", model_a_name: \\\"Child\\\", model_b_name: \\\"Parent\\\" })"""
+          // "The change you are trying to make would violate the required relation 'ChildToParent' between Child and Parent"
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideUpdateSpec.scala]: These error messages can probably be improved, should I spend time on that?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala [hunk 9]
```diff
@@ -961,8 +962,8 @@ class NestedConnectMutationInsideCreateSpec extends FlatSpec with Matchers with
          |}
       """.stripMargin,
       project,
-      errorCode = 3039,
-      errorContains = s"No Node for the model Comment with value $todoId for id found."
+      errorCode = 0, // 3039,
+      errorContains = """RecordNotFound(\"Expected 1 records to be connected, found 0.\")"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: It feels like it should be an instance of a user facing "NotFound" error.

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala [hunk 10]
```diff
@@ -48,8 +48,8 @@ class NestedConnectMutationInsideCreateSpec extends FlatSpec with Matchers with
            |}
         """.stripMargin,
         project,
-        errorCode = 3042,
-        errorContains = "The change you are trying to make would violate the required relation 'ChildToParent' between Child and Parent"
+        errorCode = 0, // 3042,
+        errorContains = """Error in query graph construction: RelationViolation(RelationViolation { relation_name: \"ChildToParent\", model_a_name: \"Child\", model_b_name: \"Parent\""""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: Should `RelationViolation` be a user-facing error, with a code? (It sounds pretty common)

File: libs/user-facing-errors/src/quaint.rs [hunk 11]
```diff
@@ -15,6 +16,7 @@ impl From<quaint::error::DatabaseConstraint> for crate::query_engine::DatabaseCo
         match other {
             quaint::error::DatabaseConstraint::Fields(fields) => Self::Fields(fields.to_vec()),
             quaint::error::DatabaseConstraint::Index(index) => Self::Index(index.to_string()),
+            _ => todo!(),
```
Comment (tomhoule) [inline libs/user-facing-errors/src/quaint.rs]: Part of a quaint upgrade that should take place in another branch.

File: query-engine/connector-test-kit/src/test/scala/queries/filters/WhereUniqueSpec.scala [hunk 12]
```diff
@@ -38,8 +38,9 @@ class WhereUniqueSpec extends FlatSpec with Matchers with ApiSpecBase {
     server.queryThatMustFail(
       s"""query{user(where: {}){unique}}""",
       project,
-      errorCode = 3040,
-      errorContains = """You provided an invalid argument for the where selector on User. Please provide exactly one unique field and value."""
+      errorCode = 2009, // 3040,
+      errorContains = """Failed to validate the query `Error occurred during query validation & transformation:\nAssertion error: Expected object to have exactly 1 key-value pairs, got: 0 ()""",
+      // """You provided an invalid argument for the where selector on User. Please provide exactly one unique field and value."""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/queries/filters/WhereUniqueSpec.scala]: Should we try to provide something closer to the old error here?

File: query-engine/connector-test-kit/src/test/scala/queries/filters/WhereUniqueSpec.scala [hunk 13]
```diff
@@ -51,9 +52,9 @@ class WhereUniqueSpec extends FlatSpec with Matchers with ApiSpecBase {
     server.queryThatMustFail(
       s"""query{user(where: {id:"wrong", email: "test@test.com"}){unique}}""",
       project,
-      errorCode = 3045,
-      errorContains =
-        """You provided more than one field for the unique selector on User. If you want that behavior you can use the many query and combine fields with AND / OR."""
+      errorCode = 2009, // 3045,
+      errorContains = """"Failed to validate the query `Error occurred during query validation & transformation:\nAssertion error: Expected object to have exactly 1 key-value pairs, got: 2 (email, id).`"""
+        // """You provided more than one field for the unique selector on User. If you want that behavior you can use the many query and combine fields with AND / OR."""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/queries/filters/WhereUniqueSpec.scala]: Providing more than one unique criteria in a findOne seems like it should be a user facing error, wdyt?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpsertSpec.scala [hunk 14]
```diff
@@ -818,7 +821,8 @@ class NestedDisconnectMutationInsideUpsertSpec extends FlatSpec with Matchers wi
          |}
       """,
       project,
-      errorCode = 3041
+      errorCode = 0, // 3041
+      errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'5\\': RecordsNotConnected { relation_name: \\\"UserFollows\\\", parent_name: \\\"User\\\", child_name: \\\"User\\\"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpsertSpec.scala]: Not sure these (in this file) should be user facing with an error code

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteMutationSpec.scala [hunk 15]
```diff
@@ -70,8 +70,8 @@ class DeleteMutationSpec extends FlatSpec with Matchers with ApiSpecBase {
     server.queryThatMustFail(
       s"""mutation {deleteScalarModel(where: {unicorn: null}){unicorn}}""",
       project = project,
-      errorCode = 3040,
-      errorContains = "You provided an invalid argument for the where selector on ScalarModel."
+      errorCode = 0, // 3040,
+      errorContains = """"Error occurred during query execution:\nInterpretationError(\"Error for binding \\'0\\': RecordNotFound(\\\"Record to delete does not exist.\\\")\""""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteMutationSpec.scala]: Seems like it should be user-facing with a code.

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/CreateMutationSpec.scala [hunk 16]
```diff
@@ -112,15 +112,16 @@ class CreateMutationSpec extends FlatSpec with Matchers with ApiSpecBase {
     server.queryThatMustFail(
       s"""mutation {createScalarModel(data: {optString: "test", optInt: B, optFloat: 1.234, optBoolean: true, optEnum: A, optDateTime: "2016-07-31T23:59:01.000Z" }){optString, optInt, optFloat, optBoolean, optEnum, optDateTime }}""",
       project = project,
-      0,
-      errorContains = """Value types mismatch. Have: Enum("B"), want: Scalar(Int)"""
+      errorCode = 0,
+      errorContains = """ObjectValidationError { object_name: \"ScalarModelCreateInput\", inner: FieldValidationError { field_name: \"optInt\", inner: ValueTypeMismatchError { have: Enum(\"B\"), want: Scalar(Int) }"""
+       // """Value types mismatch. Have: Enum("B"), want: Scalar(Int)"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/CreateMutationSpec.scala]: These should be pretty easy to make nicer

File: query-engine/connector-test-kit/src/test/scala/writes/deadlocksAndTransactions/TransactionalNestedExecutionSpec.scala [hunk 17]
```diff
@@ -214,8 +215,9 @@ class TransactionalNestedExecutionSpec extends FlatSpec with Matchers with ApiSp
          |}
       """.stripMargin,
       project,
-      errorCode = 3039,
-      errorContains = s"No Node for the model Todo with value $falseWhereInError for innerUnique found."
+      errorCode = 0, // 3039,
+      errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'1\\': AssertionError(\\\"Expected a valid parent ID to be present for nested update to-one case."""
+        // s"No Node for the model Todo with value $falseWhereInError for innerUnique found."
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/deadlocksAndTransactions/TransactionalNestedExecutionSpec.scala]: Should we turn this into a user-facing NotFound kind of error?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpdateSpec.scala [hunk 18]
```diff
@@ -939,7 +944,8 @@ class NestedDisconnectMutationInsideUpdateSpec extends FlatSpec with Matchers wi
          |}
       """,
       project,
-      errorCode = 3041
+      errorCode = 0, // 3041
+      errorContains = """Error occurred during query execution:\nInterpretationError(\"Error for binding \\'1\\': RecordsNotConnected { relation_name: \\\"UserFollows\\\", parent_name: \\\"User\\\", child_name: \\\"User\\\" }"""
```
Comment (tomhoule) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpdateSpec.scala]: The old error codes are defined here: https://github.com/prisma/prisma/blob/56ee4251864b782330865e3e8ebdf89a25e9a0da/server/connectors/api-connector/src/main/scala/com/prisma/api/schema/Errors.scala#L103

Should we try to reimplement similar handling for 3041 and 3042?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpsertSpec.scala (hunk 14 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedDisconnectMutationInsideUpsertSpec.scala]: I think they should

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedUpsertMutationInsideUpdateSpec.scala [hunk 19]
```diff
@@ -509,8 +516,8 @@ class NestedUpsertMutationInsideUpdateSpec extends FlatSpec with Matchers with A
          |}
       """,
       project,
-      errorCode = 3010,
-      errorContains = "A unique constraint would be violated on Comment. Details: Field name = uniqueComment"
+      errorCode = 2002,
+      errorContains = s"Unique constraint failed on the $errorTarget"
```
Comment (mavilein) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/NestedUpsertMutationInsideUpdateSpec.scala]: Model should be part of the error message

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala (hunk 10 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: yes

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala [hunk 20]
```diff
@@ -510,8 +511,8 @@ class NestedConnectMutationInsideCreateSpec extends FlatSpec with Matchers with
            |}
         """.stripMargin,
         project,
-        errorCode = 3039,
-        errorContains = "No Node for the model Child with value DOES NOT EXIST for c found."
+        errorCode = 0, // 3039,
+        errorContains = "RecordNotFound(\\\"Expected 2 records to be connected, found 1.\\\")"
```
Comment (mavilein) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: should be a `RecordsNotConnected` imo

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala [hunk 21]
```diff
@@ -871,8 +872,8 @@ class NestedConnectMutationInsideCreateSpec extends FlatSpec with Matchers with
          |}
       """.stripMargin,
       project,
-      errorCode = 3039,
-      errorContains = "No Node for the model Comment with value 5beea4aa6183dd734b2dbd9b for id found."
+      errorCode = 0, // 3039,
+      errorContains = "RecordNotFound(\\\"Expected 1 records to be connected, found 0.\\\")"
```
Comment (mavilein) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: should be user facing imo

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala (hunk 9 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideCreateSpec.scala]: yes :+1

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideUpdateSpec.scala (hunk 8 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedConnectMutationInsideUpdateSpec.scala]: Yes. That would be great :pray

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedDeleteMutationInsideUpdateSpec.scala [hunk 22]
```diff
@@ -225,7 +225,8 @@ class NestedDeleteMutationInsideUpdateSpec extends FlatSpec with Matchers with A
          |}
       """,
         project,
-        errorCode = 3041
+        errorCode = 0,
+        errorContains = """[Query Graph] Expected a valid parent ID to be present for a nested delete on a one-to-many relation."""
```
Comment (mavilein) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/alreadyConverted/NestedDeleteMutationInsideUpdateSpec.scala]: should be improved

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala [hunk 23]
```diff
@@ -499,8 +500,8 @@ class NestedUpdateMutationInsideUpdateSpec extends FlatSpec with Matchers with A
          |}
       """,
       project,
-      errorCode = 3040,
-      errorContains = "You provided a null value for the where clause on Todo."
+      errorCode = 2009, // 3040,
+      errorContains = "Failed to validate the query `Error occurred during query validation & transformation:\\nAssertion error: You provided a null value for a where clause (or implicit nested selector). Please provide a non null value.."
```
Comment (mavilein) [inline query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala]: Can we give details on the model here?

File: query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala (hunk 6 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/nestedMutations/notUsingSchemaBase/NestedUpdateMutationInsideUpdateSpec.scala]: yes

File: query-engine/connector-test-kit/src/test/scala/writes/scalarLists/CreateMutationListSpec.scala (hunk 7 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/scalarLists/CreateMutationListSpec.scala]: sounds right :+1

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/CreateMutationSpec.scala (hunk 16 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/CreateMutationSpec.scala]: :+1

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteMutationSpec.scala (hunk 15 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteMutationSpec.scala]: 👍

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteSpec.scala (hunk 5 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/DeleteSpec.scala]: yes

File: query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/UpdateMutationSpec.scala (hunk 1 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/topLevelMutations/UpdateMutationSpec.scala]: yes

File: query-engine/connector-test-kit/src/test/scala/writes/withId/ExecuteRawSpec.scala (hunk 2 above)
Comment (mavilein) [reply query-engine/connector-test-kit/src/test/scala/writes/withId/ExecuteRawSpec.scala]: That would be really nice

File: query-engine/connector-test-kit/src/test/scala/util/PlayJson.scala [hunk 24]
```diff
@@ -118,10 +118,11 @@ trait PlayJsonExtensions extends JsonUtils {
       val errors = json.pathAsSeq("errors")
       require(requirement = errors.size == errorCount, message = s"expected exactly $errorCount errors, but got ${errors.size} instead.")
 
+
       if (errorCode != 0) {
-        val errorCodeInResult = errors.head.pathAsLong("code")
+        val errorCodeInResult = errors.head.pathAsString("user_facing_error.error_code").stripPrefix("P")
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply query-engine/connector-test-kit/src/test/scala/util/PlayJson.scala]: Yes, it was an oversight, and now the frontend is expecting snake case there.

File: query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala [hunk 25]
```diff
@@ -22,13 +22,18 @@ class BringYourOwnIdSpec extends FlatSpec with Matchers with ApiSpecBase with Sc
...
+
       server.queryThatMustFail(
         s"""mutation {
          |  createParent(data: {p: "Parent2", id: "Own Id"}){p, id}
          |}""",
         project = project,
-        errorCode = 3010,
-        errorContains = "A unique constraint would be violated on Parent. Details: Field name = id"
+        errorCode = 2002,
+        errorContains = s"Unique constraint failed on the $errorTarget"
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply query-engine/connector-test-kit/src/test/scala/writes/ids/BringYourOwnIdSpec.scala]: Not at the quaint level, but maybe we can reinject that information in the core/sql connector

**Comment by tomhoule** [comment]
I think the next step would be splitting up the different kinds of interpretation errors (code 2016) and query graph construction errors (code 2019). However I don't want that branch to stay unmerged for too long, what do you thinK?

File: libs/user-facing-errors/src/query_engine.rs [hunk 26]
```diff
@@ -148,3 +148,75 @@ pub struct RawQueryFailed {
     pub code: String,
     pub message: String,
 }
+
+#[derive(Debug, UserFacingError, Serialize)]
+#[user_facing(code = "P2011", message = "Null constraint violation on the ${constraint}")]
+pub struct NullConstraintViolation {
+    pub constraint: DatabaseConstraint,
+}
+
```
Comment (do4gr) [inline libs/user-facing-errors/src/query_engine.rs]: Can the user tell which model /field was affected from the constraint?

Comment (tomhoule) [reply libs/user-facing-errors/src/query_engine.rs]: Sadly no, I haven't managed to extract more context from the error, and a discussion. I talked about it with Dom and it's pretty complicated, probably something we should do in another PR, if I understood right.

**Comment by tomhoule** [comment]
agreed, there's lots of room for improvement, but it's hard.

**Pull request merged**